File: src/main/java/me/lordsaad/modeoff/api/PlotManager.java
package me.lordsaad.modeoff.api;

import me.lordsaad.modeoff.Modeoff;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.UUID;

/**
 * Created by LordSaad.
 */
public class PlotManager {

	public UUID uuid;
	@Nullable
	public EntityPlayer player;
	public int plotID = -1;
	public BlockPos corner1, corner2;

	public PlotManager(@NotNull EntityPlayer player) {
		this.uuid = player.getUniqueID();
		this.player = player;

		PlotAssigningManager manager = PlotAssigningManager.INSTANCE;
		plotID = manager.getPlotForUUID(uuid);
		initPlot();
	}

	public PlotManager(World world, int plotID) {
		this.plotID = plotID;
		initPlot();

		UUID uuid = PlotAssigningManager.INSTANCE.getUUIDForPlot(plotID);
		if (uuid == null) {
			Modeoff.logger.error("uuid for plot " + plotID + " could not be found.");
			return;
		}
		this.uuid = uuid;

		player = world.getPlayerEntityByUUID(uuid);
	}

	public static void teleportToPlot(EntityPlayer player, int plotID) {
		if (plotID < 0) return;
		if (player == null) return;

		if (player.world.provider.getDimension() != ConfigValues.plotWorldDimensionID)
			player.changeDimension(ConfigValues.plotWorldDimensionID);

		BlockPos pos = getPlotPos(plotID);
		if (pos == null) return;

		player.setPositionAndUpdate(pos.getX() + 0.5, pos.getY() + 2, pos.getZ() + 0.5);
	}

	@Nullable
	public static BlockPos getPlotPos(int plotID) {
		if (plotID < 0) return null;

		BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos(ConfigValues.x, ConfigValues.y, ConfigValues.z);
		pos.add(ConfigValues.plotSize / 2, ConfigValues.plotSize / 2, ConfigValues.plotSize / 2);

		int row = plotID / ConfigValues.gridRows;
		int column = plotID % ConfigValues.gridColumns;

		pos.move(EnumFacing.valueOf(ConfigValues.directionOfRows), row * ConfigValues.plotSize + row * ConfigValues.plotMarginWidth + ConfigValues.plotSize / 2);
		pos.move(EnumFacing.valueOf(ConfigValues.directionOfColumns), column * ConfigValues.plotSize + column * ConfigValues.plotMarginWidth + ConfigValues.plotSize / 2);
		return new BlockPos(pos);
	}

	private void initPlot() {
		BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos(ConfigValues.x, ConfigValues.y, ConfigValues.z);
		pos.add(ConfigValues.plotSize / 2, ConfigValues.plotSize / 2, ConfigValues.plotSize / 2);

		int row = plotID / ConfigValues.gridRows;
		int column = plotID % ConfigValues.gridColumns;

		pos.move(EnumFacing.valueOf(ConfigValues.directionOfRows), row * ConfigValues.plotSize + row * ConfigValues.plotMarginWidth);
		pos.move(EnumFacing.valueOf(ConfigValues.directionOfColumns), column * ConfigValues.plotSize + column * ConfigValues.plotMarginWidth);
		corner1 = new BlockPos(pos.getX() - (ConfigValues.plotSize / 2), pos.getY(), pos.getZ() - (ConfigValues.plotSize / 2));
		corner2 = new BlockPos(pos.getX() + (ConfigValues.plotSize / 2), pos.getY(), pos.getZ() + (ConfigValues.plotSize / 2));
	}

	public void teleportToCenter() {
		teleportToPlot(player, plotID);
	}

}

File: src/main/kotlin/me/lordsaad/modeoff/api/PlotManager.kt
package me.lordsaad.modeoff.api

import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos

import java.util.UUID

/**
 * Created by LordSaad.
 */
class PlotManager {
    companion object {
        fun teleportToPlot(player: EntityPlayer?, plotID: Int) {
            if (plotID < 0) return
            if (player == null) return

            if (player.world.provider.dimension != ConfigValues.plotWorldDimensionID)
                player.changeDimension(ConfigValues.plotWorldDimensionID)

            val pos = getPlotPos(plotID) ?: return

            player.setPositionAndUpdate(pos.x + 0.5, (pos.y + 2).toDouble(), pos.z + 0.5)
        }

        fun getPlotPos(plotID: Int): BlockPos? {
            if (plotID < 0) return null

            val pos = BlockPos.MutableBlockPos(ConfigValues.firstPlotX, ConfigValues.firstPlotY, ConfigValues.firstPlotZ)
            pos.add(ConfigValues.plotSize / 2, ConfigValues.plotSize / 2, ConfigValues.plotSize / 2)

            val row = plotID / ConfigValues.plotGridRows
            val column = plotID % ConfigValues.plotGridColumns

            pos.move(EnumFacing.valueOf(ConfigValues.directionOfRows), row * ConfigValues.plotSize + row * ConfigValues.plotMarginWidth + ConfigValues.plotSize / 2)
            pos.move(EnumFacing.valueOf(ConfigValues.directionOfColumns), column * ConfigValues.plotSize + column * ConfigValues.plotMarginWidth + ConfigValues.plotSize / 2)
            return BlockPos(pos)
        }
    }


    val uuid: UUID
    var plotID = -1
    val corner1: BlockPos
    val corner2: BlockPos

    constructor(uuid: UUID, plotId: Int) {
        this.uuid = uuid
        this.plotID = plotId

        val pos = BlockPos.MutableBlockPos(ConfigValues.firstPlotX, ConfigValues.firstPlotY, ConfigValues.firstPlotZ)
        pos.add(ConfigValues.plotSize / 2, ConfigValues.plotSize / 2, ConfigValues.plotSize / 2)

        val row = plotID / ConfigValues.plotGridRows
        val column = plotID % ConfigValues.plotGridColumns

        pos.move(EnumFacing.valueOf(ConfigValues.directionOfRows), row * ConfigValues.plotSize + row * ConfigValues.plotMarginWidth)
        pos.move(EnumFacing.valueOf(ConfigValues.directionOfColumns), column * ConfigValues.plotSize + column * ConfigValues.plotMarginWidth)
        corner1 = BlockPos(pos.x - ConfigValues.plotSize / 2, pos.y, pos.z - ConfigValues.plotSize / 2)
        corner2 = BlockPos(pos.x + ConfigValues.plotSize / 2, pos.y, pos.z + ConfigValues.plotSize / 2)
    }

    constructor(player: EntityPlayer) : this(player.uniqueID, PlotAssigningManager.getPlotForUUID(player.uniqueID))

    fun teleportPlayerToCenter(player: EntityPlayer) {
        teleportToPlot(player, plotID)
    }
}
